vdyp-core/src/main/java/ca/bc/gov/nrs/vdyp/common/ValueOrMarker.java

```diff
@@ -0,0 +1,123 @@
+package ca.bc.gov.nrs.vdyp.common;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Type union of a value and an out of band marker.
+ *
+ * @author Kevin Smith, Vivid Solutions
+ *
+ * @param <Value>
+ * @param <Marker>
+ */
+public class ValueOrMarker<Value, Marker> {
+	private final boolean isMarker;
+	private final Object obj;
+
+	private ValueOrMarker(Object obj, boolean isMarker) {
+		Objects.requireNonNull(obj, "The Value or Marker of a ValueOrMarker must not be null");
+
+		this.isMarker = isMarker;
+		this.obj = obj;
+	}
+
+	/**
+	 * Is there a value
+	 *
+	 * @return
+	 */
+	public boolean isMarker() {
+		return isMarker;
+	}
+
+	/**
+	 * Is there a marker
+	 *
+	 * @return
+	 */
+	public boolean isValue() {
+		return !isMarker;
+	}
+
+	/**
+	 * Get the value if present
+	 *
+	 * @return
+	 */
+	@SuppressWarnings("unchecked")
+	public Optional<Value> getValue() {
+		if (isValue()) {
+			return (Optional<Value>) Optional.of(obj);
+		}
+		return Optional.empty();
+	}
+
+	/**
+	 * Get the marker if present
+	 *
+	 * @return
+	 */
+	@SuppressWarnings("unchecked")
+	public Optional<Marker> getMarker() {
+		if (isMarker()) {
```

Comment:
Shouldn't this throw when it's not a Marker?

Reply:
The value and the marker are `Optional` to represent this. The calling code can `.orElseThrow(...)` to throw an exception or choose to handle it some other way.   A `ValueOrMarker` will always have one of them present and the other empty.